src/Resolver/Asset/ResourceAssetResolverChain.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Atoolo\GraphQL\Search\Resolver\Asset;
+
+use Atoolo\GraphQL\Search\Resolver\Resolver;
+use Atoolo\GraphQL\Search\Types\Asset;
+use Atoolo\Resource\Resource;
+use Overblog\GraphQLBundle\Definition\ArgumentInterface;
+
+class ResourceAssetResolverChain implements Resolver
+{
+    /**
+     * @var array<string, ResourceAssetResolver>
+     */
+    private readonly array $regularAssetResolvers;
+
+    /**
+     * @var array<string, ResourceAssetResolver>
+     */
+    private readonly array $symbolicAssetResolvers;
+
+    /**
+     * @param array<string, ResourceAssetResolver> $regularAssetResolvers
+     * @param array<string, ResourceAssetResolver> $symbolicAssetResolvers
+     */
+    public function __construct(
+        iterable $regularAssetResolvers,
+        iterable $symbolicAssetResolvers
+    ) {
+        $this->regularAssetResolvers =
+            $regularAssetResolvers instanceof \Traversable ?
+            iterator_to_array($regularAssetResolvers) :
+            $regularAssetResolvers;
+        $this->symbolicAssetResolvers =
+            $symbolicAssetResolvers instanceof \Traversable ?
+            iterator_to_array($symbolicAssetResolvers) :
+            $symbolicAssetResolvers;
+    }
+
+    /**
+     * Iterates over all regular asset resolvers (tagged with
+     * `atoolo_graphql_search.resolver.asset.regular`) and all
+     * symbolic asset resolvers (tagged with
+     * `atoolo_graphql_search.resolver.asset.symbolic`) and returns
+     * the first resolved asset that is not null. All  regular asset
+     * resolvers will always have priority over all symbolic asset resolvers.
+     * Passing the argument `forceSymbolic` will cause this method to
+     * only apply symbolic asset resolvers, ignoring the regular assert
+     * resolvers.
+     */
+    public function getAsset(
+        Resource $resource,
+        ArgumentInterface $args
+    ): ?Asset {
+        if ($args['forceSymbolic'] ?? false) {
+            return $this->getAssetSymbolic($resource, $args);
+        }
+        return $this->getAssetRegular($resource, $args)
+            ?? $this->getAssetSymbolic($resource, $args);
```

Comment:
That's what I thought as well. But if we only use one tagged iterator, how would the parameter `forceSymbolic` work then? The `ResourceAssetResolverChain` needs a way to tell "regular" resolvers apart from "symbolic" resolvers in order to figure out what resolvers to apply when `forceSymbolic` is set to true.

> Theoretically, there could also be a regular resolver that should come after the symbolic link.

Should we even allow such case? I thought by concept the symbolic images  _always_ come as a fallback after all "regular" resolvers have failed.

Reply:
I just meant that:

```yaml
services:
    Atoolo\GraphQL\Search\Resolver\Asset\ResourceSymbolicImageResolver:
      arguments:
        - '@atoolo_graphql_search.resolver.url_rewriter'
        - '@atoolo_resource.navigation_hierarchy_loader'
      tags:
        - { name: 'atoolo_graphql_search.resolver.asset.symbolic', priority: 10 }
        - { name: 'atoolo_graphql_search.resolver.asset.regular', priority: 5 }
```

Then that should be enough:

```php
        if ($args['forceSymbolic'] ?? false) {
            return $this->getAssetSymbolic($resource, $args);
        }
        return $this->getAssetRegular($resource, $args);
```

Otherwise you could [hereby](https://symfony.com/doc/current/service_container/tags.html#tagged-services-with-index) reconsider whether this helps. But I would only do this if it needs to be extended. That's what I needed for the[ `DelegatingTeaserFactory`](https://github.com/sitepark/atoolo-graphql-search-bundle/blob/main/src/Resolver/DelegatingTeaserFactory.php).

Reply:
Then you could perhaps use [`index_by`](https://symfony.com/doc/current/service_container/tags.html#the-index-by-indexattribute-option) and do it that way:

```php
    public function getAsset(
        Resource $resource,
        ArgumentInterface $args
    ): ?Asset {
        $resolveOrder = $args['resolveOrder'] ?? array_keys($this->assetResolvers);
        foreach ($resolveOrder  as $resolverKey) {
            $assetResolver = $this->assetResolvers[$resolverKey];
            $asset = $assetResolver->getAsset(
                $resource,
                $args
            );
            if ($asset !== null) {
                return $asset;
            }
        }
        return null;
    }
```

In the GraphQL query, you would then not specify `forceSymbolic:true`, but `resolveOrder:["symbolicImage"]`.

The services would then be configured as follows:

```yaml
services:
    Atoolo\GraphQL\Search\Resolver\Asset\ResourceSymbolicImageResolver:
      arguments:
        - '@atoolo_graphql_search.resolver.url_rewriter'
        - '@atoolo_resource.navigation_hierarchy_loader'
      tags:
        - { name: 'atoolo_graphql_search.resolver.asset', priority: 5, key: 'symbolicImage' }
```

But I don't know whether `key` and `priority` can be combined.

Reply:
Hm, I think we had this discussion before :smile:
Didn't we commit to the idea that `asset` shouldn't have a string parameter that represents a type (or in this case a "resolveOrder", which would have the same problems)? 

Maaaybe we should do it as follows: The client shouldn't have any control over what `asset` resolves to, so there's no need for a new parameter. It's entirely up to the backend what `asset` returns. The `getAsset` function will simply iterate over all asset resolvers and will not inherently know which resolvers are "symbolic" and which aren't. Any fallback mechanics regarding symbolic images can then be solved by managing the resolvers priorities.

However, since symbolic images seem to be a somewhat special attribute for teasers, maybe we should actually just add a new optional field `symbolicImage` to the teaser type.  We previously argued that this seems to be contradictory to the concept of the `asset` field, but I would argue that symbolic images are actually a special kind of asset for teasers that is not comparable to any other asset type.

So a teaser would look like this:
```graphql
type ArticleTeaser implements Teaser {
  ...
  asset(variant: String!): Asset
  symbolicImage: SymbolicImage
}
```

For most use cases, the client can simply leave out the `symbolicImage` field, since the `asset` field can be trusted to resolve to the "correct" Asset. For any other use case, the client can then use the `symbolicImage` field instead of the `asset` field. 
The client would simply have to write another query or use [directives](https://graphql.org/learn/queries/#directives) to conditionally request the `symbolicImage` field.

Reply:
Yes, I think that's better.